Create GRADING.md (#3)
cbfacademy-solutions/java-api-assessment: GRADING.md

> # Assessment Grading Guide for API Development with Java 17 & SpringBoot

Reviewer: Spring Boot

Welcome to the grading guide for the API Development Assessment. This guide has been meticulously designed to ensure a clear, consistent, and comprehensive evaluation of students' projects. As an assessor, you'll find a rubric below that breaks down the evaluation criteria into four distinct sections: 
- Completion
- Functionality
- Code Quality & Structure
- Language Profiency 

Please go through each section methodically, ensuring that you understand the criteria and can confidently attribute a score from 0 to 5 for each student's work based on their performance in the listed areas.

## Table of Contents

1. [Completion](#completion)
1. [Functionality](#functionality)
1. [Code Quality & Structure](#code-quality--structure)

> 1. [Language Proficiency](#language-proficiency)

Reviewer: Just noticed that these are all numbered 1

Reviewer 2: That's valid syntax for [numbered lists](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#lists)

## Completion

This section gauges the progress of the student in terms of repository management and project submission. It's a reflection of commitment and the timely delivery of the project.

| **Score** | **Caption**                                          | **Details**                                        |
|-----------|------------------------------------------------------|----------------------------------------------------|
| 0         | No repository created                               | Student has not initialised any repository for the project.                                               |

> | 1         | Repository created, but no code commits             | Repository exists but lacks any code. This indicates no progress on the project.                           |

Reviewer 3: ```suggestion
| 1         | Repository created, but has no commit history             | Repository exists but has no commits. This indicates no progress on the project.                           |
```

| 2         | Completed late without agreed extension             | Project was submitted after the deadline without any prior agreement for extension.                        |
| 3         | Completed within agreed extension                   | Project was submitted late, but within an agreed-upon extension timeframe.                                 |
| 4         | Completed within deadline                           | Project was submitted on time, indicating good time management.                                           |
| 5         | Completed within deadline, with full commit history | Project was submitted on time and showcases a consistent work pattern with a well-maintained commit history.|

### Tips

- **Repository Check**: Verify the existence of the repository and its contents to confirm the student's engagement with the project.
- **Commit Analysis**: Examine the commit history for patterns of work and consistency, reflecting the student's workflow and time management.

> - **Deadline Adherence**: Assess whether the project was completed on time or if an extension was justified and used effectively.

Reviewer: Should this be a task for the assessor? Or would they just refer to the table and see whether it indicates that an extension was agreed? 

Would suggest this says: 
- **Deadline Adherence**: Verify whether the project was completed on time or within an agreed deadline

| 1         | Basic API with significant errors                    | The API has some functionality but is plagued by errors and does not meet all requirements.                                                                                                        |
| 2         | API mostly functional, missing some key requirements | For instance, the algorithm might be implemented but no unit testing is done.                                                                                                                      |
| 3         | API functional, meets most requirements              | Only one or two of the listed criteria (like CRUD operations or exception handling) are missing or poorly executed.                                                                                |
| 4         | API fully functional, minor improvements needed      | All criteria are met but some areas, such as documentation or HTTP protocol usage, could use refinement.                                                                                          |

> | 5         | API is robust, user-friendly, and meets all requirements | All criteria, from algorithm implementation to good documentation, are perfectly executed.                                                                                                          |

Reviewer 3: For this criteria, we've usually reserved top marks for students that go above and beyond the requirements. So I'd suggest the following amendments:

```suggestion
| **Score** | **Caption**                                                                     | **Details**                                                                                              |
| --------- | ------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------- |
| 0         | No working functionality                                                        | The API has no operational endpoints.                                                                    |
| 1         | API is only partially functional                                                | The API has limited functionality and only meets a few requirements.                                     |
| 2         | API is mostly functional, but missing some key requirements                     | For instance, the algorithm might be implemented but no unit testing is done.                            |
| 3         | API is fully functional, but minor improvements needed                          | All criteria are met but some areas, such as documentation or HTTP protocol usage, could use refinement. |
| 4         | API is robust, user-friendly, and meets all requirements                        | All criteria, from algorithm implementation to good documentation, are perfectly executed.               |
| 5         | API fully meets all requirements and also demonstrates additional functionality | All criteria are met and the solution provides extra utility.                                            |
```

Removed references to errors as that should be addressed by Code Quality metric.

### Tips

- **Endpoint Testing**: Test each API endpoint to ensure they are working as expected and meet the project specifications.
- **User Experience**: Consider the ease of use and whether the API provides helpful feedback during interaction.
- **Technical Requirements**: Check for the implementation of all the technical requirements listed in the project brief.

---

> ## Code Quality & Structure
> 
> This section assesses the integrity and organisation of the code. It evaluates readability, maintainability, efficiency, modularity, and adherence to Java + Spring Boot coding standards.
> 
> | **Score** | **Caption**                                                | **Details**                                                                                                                            |
> |-----------|------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
> | 0         | No adherence to guidelines or organisation                 | Code lacks structure and organisation as per Java + Spring Boot recommendations, with everything in one file or lacking clear sections. |
> | 1         | Minimal adherence, multiple errors, and organisation       | Code has basic structure and organisation but with multiple errors, lacking in modularity, inheritance, and proper exception handling. |
> | 2         | Decent structure, needs optimisation and better structure  | Organised code with clear separation of concerns but requires significant improvement in algorithm, unit testing, and modularity.      |
> | 3         | Good overall but needs refactoring and better documentation | Mostly consistent structure with Java + Spring Boot standards but needs refining and additional documentation.                          |
> | 4         | Very good, minor improvements in structure needed          | Nearly aligns perfectly with guidelines, showing clear modularity and organisation but may have minor areas for improvement.           |
> | 5         | Perfectly structured and organised as per guidelines       | Exemplary code that aligns perfectly with Java + Spring Boot recommendations, demonstrating clear modularity and comprehensive documentation. |

Reviewer 3: ```suggestion
## Solution Quality

This section assesses the integrity and organisation of the code. It evaluates readability, maintainability, efficiency, modularity, and adherence to Java and Spring Boot coding standards.

| **Score** | **Caption**                                                    | **Details**                                                                                                                                            |
| --------- | -------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| 0         | Solution does not compile                                      | Code is missing, incomplete or contains compile-time errors that prevent it from building.                                                             |
| 1         | Solution compiles, but with warnings                           | IDE and/or compiler raises multiple warnings, indicating highly sub-optimal implementation.                                                            |
| 2         | Solution compiles successfully, but is poorly structured       | Code is free of errors or warnings, but is hard to navigate, doesn't use packages effectively, etc.                                                    |
| 3         | Solution is reasonably structured, but in need of refactoring  | Code is fairly organised but has overweight classes/methods, inconsistent/unintuitive naming conventions, etc.                                         |
| 4         | Solution is well structured, but has some room for improvement | Code demonstrates principles like DRY, SRP, encapsulation and clear naming conventions, but could have better testing, formatting, documentation, etc. |
| 5         | Solution consistently demonstrates best practices              | Exemplary code that closely follows recommended Java and Spring Boot conventions, with thorough test coverage and comprehensive documentation.         |
```

### Tips

- **Readability and Maintainability**: Review the code for clear naming conventions, logical structuring, and the inclusion of comments where necessary.
- **Efficiency**: Evaluate if the code is not only functional but also written in a way that is efficient and optimised for performance.

> - **Best Practices**: Ensure that the code adheres to Java and SpringBoot best practices, including the use of proper annotations and dependency management.

Reviewer: Spring Boot

---

> ## Language Proficiency Rubric for Assessors
> 
> This section evaluates students' Java coding capabilities, focusing on the robustness of exception handling and the effectiveness of data structure usage.
> 
> | **Score** | **Caption**                                    | **Details for Assessors**                                                                                                                                                 |
> |-----------|------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
> | 0         | Non-Functional                                 | The code does not compile or run. There is no evidence of attempted exception handling or data structure usage. Assess if there is an attempt to complete the task.       |
> | 1         | Basic Functionality                            | The code compiles and runs but with minimal or ineffective exception handling. Basic data structures are used, indicating only a rudimentary understanding.               |
> | 2         | Intermediate Proficiency                       | The code runs with basic exception handling; common exceptions may be caught. Data structures are used correctly but not optimally.                                       |
> | 3         | Competent Exception Handling and Data Usage    | The code shows competent use of exception handling including some custom exceptions. Data structures are effectively used but may not be the most efficient choice.       |
> | 4         | Advanced Coding and Structure Understanding    | Exception handling is robust with clear user feedback. Data structures are well-chosen and demonstrate a good understanding of Java's Collections Framework.             |
> | 5         | Exceptional Skill and Understanding            | The code reflects exceptional skill in Java. Exception handling is comprehensive, covering all possible edge cases. Data structures are utilised optimally, showcasing excellent command over Java language features. |

Reviewer 3: ```suggestion
## Language Proficiency

This section evaluates students' Java coding capabilities, focussing on the robustness of exception handling and the effectiveness of data structure usage.

| **Score** | **Caption**                 | **Details for Assessors**                                                                                                                                                                        |
| --------- | --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| 0         | No Demonstrable Proficiency | No code has been submitted for evaluation.                                                                                                                                                       |
| 1         | Minimal Proficiency         | Ignores errors/warnings, provides no effective exception handling, only basic data structures and control flows used, indicating only a rudimentary understanding.                                      |
| 2         | Basic Proficiency           | Implements basic error handling with simple try-catch blocks and uses standard data structures. However, there's limited usage of built-in Java packages.           |
| 3         | Intermediate Proficiency    | Includes competent error handling, utilises efficient data structures and demonstrates good Java coding practices, making appropriate use of platform packages.         |
| 4         | Advanced Proficiency        | Exhibits advanced exception handling techniques like custom exceptions, with effective use of third-party packages. Demonstrates deep understanding of Java concepts and efficient coding. |
| 5         | Exceptional Proficiency     | Reflects exemplary skill in Java, showcasing an excellent command of advanced Java language features and demonstrable research/practise beyond the scope of course material.                    |
```
